Remove Parse::Node, add ElementIndex in docs for typed insts

danakj opened the pull request:
The documentation still referred to typed instructions having a Parse::Node field, however that was removed and moved to the InstStore in f197219c107644.

Then GetParseNode() was renamed to GetNodeId() in 86a7c9ff451f4cf and then GetLocationId() in b079acd86f00e593 and finally GetLocId() in b5d28f2c4b0fee46.

The comment in typed_inst.h mentions only three fields now, but some types still have four, thanks to the unmentioned `ElementIndex index` field. Normally this field comes last, after the `[...]Id` fields except for in one case, AssociatedEntity. Rather than write ambiguously ordered documentation, update the comment to and docs to say that the ElementIndex comes last, and move it to the last position in AssociatedEntity. Tests are rebased accordingly.

danakj commented:
I know there's some StructReflection which in theory could be made to do the wrong thing by changing field orders, but I couldn't find anything like that affected here, but just in case pointing it out?

danakj committed:
typos

danakj committed:
more-typo

zygoloid commented on toolchain/docs/check.md:
```diff
@@ -96,10 +92,21 @@ instruction, as described in
 -   Up to two additional, kind-specific members. For example `SemIR::Assign` has
     members `InstId lhs_id` and `InstId rhs_id`.
 
+-   Optionally, an `ElementIndex index` field for instructions that refer to a
```
This is one of the kind-specific members described in the previous bullet, rather than a distinct special case. I don't think it's worth calling out as a special case here, but maybe the previous bullet could mention there are various different valid types for the kind-specific members?

danakj replied:
I've combined these two bullet points together, using them as two examples.

zygoloid commented on toolchain/sem_ir/typed_insts.h:
```diff
@@ -21,6 +21,8 @@
 //   value. This includes instructions that produce an abstract value, such as a
 //   `Namespace`, for which a placeholder type should be used.
 // - Up to two `[...]Id` members describing the contents of the struct.
```
Maybe we should say: up to two members of types derived from `IdBase` rather than inaccurately saying `` `[...]Id` ``.

zygoloid replied:
Though even that's not quite right. `IntId` and `BuiltinInstKind` are the exceptions. The real rule is that they need to be listed in the `IdKind` enum and implement the `FromRaw`/`ToRaw` protocol in `inst.h`.

danakj replied:
Ah okay, also there's one type with three IdBase members:
```cc
struct AnyAggregateAccess {
  static constexpr InstKind Kinds[] = {
      InstKind::StructAccess, InstKind::TupleAccess,
      InstKind::ClassElementAccess, InstKind::InterfaceWitnessAccess};

  InstKind kind;
  TypeId type_id;
  InstId aggregate_id;
  ElementIndex index;
};```

danakj replied:
Ah but one is type_id, okay nvm. This one threw me off but that was because I forgot the type_id rule.

danakj replied:
A couple more observations though:
- `AnyFoundationDecl` and `AnyBranch` have an `int32_t` member, rather than something like an `ElementIndex`. `ToRaw(int32_t)` does not appear to be defined.
- While `IntId` is, `BuiltinInstKind` is _not_ in the `IdKind` enum. But the latter is also not a field in any typed inst. Rather than having a `BuiltinInst` typed inst, we now have a separate struct for each one, introduced in be5db6e1cdf6a815f3.
- The idea that builtins don't have a location id comes from the old text around `SemIR::Builtin`, a type that no longer exists. I can't find a reason why this would be true in the code, the parsing adds a TokenIndex at the location of a builtin instruction that should come with a source location. I am guessing this was maybe referring to the _definition_ of builtins? And that this just no longer exists, as they are proper builtin instructions now? As such I think I would remove that clause and just say all instructions come with a location?

danakj replied:
Ok I understand the first two points now, and I think we can make this more explicit in the code and do some cleanup as a result, which I've done in https://github.com/carbon-language/carbon-lang/pull/4606

danakj committed:
reviewer feedback

danakj committed:
put-field-order-back

danakj committed:
Remove Parse::Node, better describe fields in docs for typed insts

The documentation still referred to typed instructions having a
Parse::Node field, however that was removed and moved to the InstStore
in f197219c107644.

Then GetParseNode() was renamed to GetNodeId() in 86a7c9ff451f4cf and
then GetLocationId() in b079acd86f00e593 and finally GetLocId() in
b5d28f2c4b0fee46.

The comment in typed_inst.h and docs in check.md don't fully explain
what can be used as a field type in the typed instruction structs.
Update the docs to explain that any type in `SemIR::IdKind` can be used
as a field. This includes ElementIndex, which does not end in an `...Id`
suffix like the current documented rule requires.

danakj committed:
more-tweaks-to-md

zygoloid commented on toolchain/docs/check.md:
```diff
@@ -84,37 +84,42 @@ instruction, as described in
+A number of instruction types in
+[sem_ir/typed_insts.h](/toolchain/sem_ir/typed_insts.h) are builtin
+instructions, such as `SemIR::TypeType` which represents the unconstrained facet
+type `type`. We will also have built-in functions which would need to form the
+implementation of some library types, such as `i32`. Built-ins have stable ids
+in the `SemIR::InstStore` across `SemIR` instances.
```
```suggestion
type `type`. Builtins have stable ids in the `SemIR::InstStore` across `SemIR`
instances.
```
Pre-existing, but since we're here: I think mentioning builtin functions in this paragraph is a bit misleading, since they don't really have any connection to builtin instructions. We should probably have some separate description of them elsewhere in this document, but I think they deserve a section of their own.

danakj committed:
reviewer-suggestion

Co-authored-by: Richard Smith <richard@metafoo.co.uk>

danakj commented:
Thanks!
